CMakeLists.txt

```diff
@@ -0,0 +1,969 @@
+check_include_files( "stdint.h" HAVE_STDINT_H )
+check_include_files( "sys/types.h" HAVE_SYS_TYPES_H )
+
+include( CheckIncludeFileCXX )
+check_include_file_cxx( "istream" HAVE_ISTREAM )
+check_include_file_cxx( "ostream" HAVE_OSTREAM )
+check_include_file_cxx( "sstream" HAVE_SSTREAM )
+
+# Check types exist
+include( CheckTypeSize )
+check_type_size( "long long" LONG_LONG_SIZE ) # also sets HAVE_LONG_LONG
+if ( LONG_LONG_SIZE GREATER 0 )
+  set( HAVE_LONG_LONG ON )
+endif ()
+
+set( CMAKE_EXTRA_INCLUDE_FILES "types.h" )
+check_type_size( u_int16_t U_INT16_T_SIZE )
+set( CMAKE_EXTRA_INCLUDE_FILES OFF )
+if ( U_INT16_T_SIZE GREATER 0 )
+  set( HAVE_U_INT16_T ON )
+endif ()
+
+set( CMAKE_EXTRA_INCLUDE_FILES "stdint.h" )
+check_type_size( uint16_t UINT16_T_SIZE )
+set( CMAKE_EXTRA_INCLUDE_FILES OFF )
+if ( UINT16_T_SIZE GREATER 0 )
+  set( HAVE_UINT16_T ON )
+endif ()
+
+# Check symbols / defines exist
+include( CheckSymbolExists )
+check_symbol_exists( NAN "math.h" HAVE_NAN )
+check_symbol_exists( isnan "math.h" HAVE_ISNAN )
+
+include( CheckCXXSymbolExists )
+check_cxx_symbol_exists( M_E "cmath" HAVE_M_E )
+check_cxx_symbol_exists( M_PI "cmath" HAVE_M_PI )
+
+# Check functions exist
+include( CheckFunctionExists )
+check_function_exists( expm1 "math.h" HAVE_EXPM1 )
+
+################################################################################
+##################           Find utility programs            ##################
+################################################################################
+
+# needed for target doc and fulldoc
+find_package( Doxygen )
+find_program( SED NAMES sed gsed )
+
+################################################################################
+##################          Process custom options            ##################
+################################################################################
+
+# add custom warnings and optimizations
+if ( with-optimize )
+  if ( with-optimize STREQUAL "ON" )
+    set( with-optimize "-O3" )
+  endif ()
+  foreach ( flag ${with-optimize} )
+    set( CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${flag}" )
+    set( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${flag}" )
+  endforeach ()
+endif ()
+
+if ( with-debug )
+  if ( with-debug STREQUAL "ON" )
+    set( with-debug "-g" )
+  endif ()
+  foreach ( flag ${with-debug} )
+    set( CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${flag}" )
+    set( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${flag}" )
+  endforeach ()
+endif ()
+
+if ( with-warning )
+  if ( with-warning STREQUAL "ON" )
+    set( with-warning "-Wall" )
+  endif ()
+  foreach ( flag ${with-warning} )
+    set( CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${flag}" )
+    set( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${flag}" )
+  endforeach ()
+endif ()
+
+if ( with-libraries )
+  if ( with-libraries STREQUAL "ON" )
+    message(FATAL_ERROR "-Dwith-libraries requires full library paths.")
+  endif ()
+  foreach ( lib ${with-libraries} )
+    if ( EXISTS "${lib}" )
+      link_libraries( "${lib}" )
```

Comment:
Eclipse tells me that `link_libraries` is deprecated and that one should use `target_link_libraries` instead, not sure how important this is.